### PR TITLE
BHBC-1457

### PR DESCRIPTION
**Pull request by cgarrettjones**
BHBC-1457

## Links to jira tickets

- - [BHBC-1457](https://quartech.atlassian.net/browse/BHBC-1495)

## This PR contains the following changes

- survey occurrence submission EML creation and addition to S3 DwCA

## This PR contains the following types of changes

- [ ] New feature (change which adds functionality)
- [ ] Enhancement (improvements to existing functionality)
- [ ] Bug fix (change which fixes an issue)
- [ ] Misc cleanup / Refactoring / Documentation
- [ ] Version change

## Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### General

- [ ] I have performed a self-review of my own code

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Screenshots

Please add any relevant UI screenshots, if applicable.

### CODE REVIEW

File: api/src/queries/dwc/dwc-queries.ts [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+
+const defaultLog = getLogger('queries/dwc/dwc-queries');
+
+/**
+ * SQL query to get submission occurrence record given package ID for a particular survey.
+ *
+ * @param {number} dataPackageId
+ * @returns {SQLStatement} sql query object
+ */
+export const getSurveyOccurrenceSubmissionSQL = (dataPackageId: number): SQLStatement | null => {
```
Comment (NickPhura) [inline api/src/queries/dwc/dwc-queries.ts]: Small nit: this function (and most of them I believe) dont ever return `null`.  So the return type could be updated to just say `SQLStatement`

```suggestion
export const getSurveyOccurrenceSubmissionSQL = (dataPackageId: number): SQLStatement => {
```

File: api/src/utils/db-constant-utils.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * Get db character metadata constants.
+ *
+ * @param {string} constantName
+ * @param {IDBConnection} connection
+ * @return {*} {Promise<void>}
+ */
+export const getDbCharacterSystemMetaDataConstant = async (
+  constantName: string,
+  connection: IDBConnection
+): Promise<any> => {
```
Comment (NickPhura) [inline api/src/utils/db-constant-utils.ts]: nit: this returns either a string (or number?) or null;

Something like: `Promise<string | null>`

Same for the JSDoc on line 10.

File: api/src/paths/dwc/eml.ts [hunk 3]
```diff
@@ -0,0 +1,921 @@
...
+ *
+ * @param {number} data_package_id
+ * @param {IDBConnection} connection
+ * @return {*} {Promise<void>}
+ */
+export const getDataPackage = async (dataPackageId: number, connection: IDBConnection): Promise<any> => {
+  const sqlStatement = getDataPackageSQL(dataPackageId);
+
+  if (!sqlStatement) {
+    throw new HTTP400('Failed to build SQL update statement');
```
Comment (NickPhura) [inline api/src/paths/dwc/eml.ts]: small nit: these all say `update` statement, maybe update to `get`

File: api/src/queries/dwc/dwc-queries.ts (hunk 1 above)
Comment (cgarrettjones) [reply api/src/queries/dwc/dwc-queries.ts]: adjusted

File: api/src/utils/db-constant-utils.ts (hunk 2 above)
Comment (cgarrettjones) [reply api/src/utils/db-constant-utils.ts]: adjusted

File: api/src/paths/dwc/eml.ts (hunk 3 above)
Comment (cgarrettjones) [reply api/src/paths/dwc/eml.ts]: adjusted

File: api/src/paths/dwc/eml.ts [hunk 4]
```diff
@@ -0,0 +1,918 @@
...
+    ]
+  };
+
+  if (projectFundingSource.length) {
+    emlRoot.dataset.project.relatedProject.funding = getFundingEML(projectFundingSource);
+  }
+
+  emlRoot.dataset.project.relatedProject.studyAreaDescription = { coverage: {} };
+  emlRoot.dataset.project.relatedProject.studyAreaDescription.coverage.geographicCoverage = getGeographicCoverageEML(
+    'Not Supplied',
```
Comment (NickPhura) [inline api/src/paths/dwc/eml.ts]: Is this meant to be hardcoded? 

(noticed that when I added a project boundary description it didnt show up int he eml)

**Comment by NickPhura** [comment]
A few things that I noticed are missing from the EML:

Project:
- Permits
- Location boundary description

Survey:
- Permits
- Survey Methodology
- Ancillary Species

I can't speak to if all of those items are essential or not, mostly permits stands out to me. But everything else in the project/survey meta is there, I believe.

File: api/src/paths/dwc/eml.ts (hunk 4 above)
Comment (cgarrettjones) [reply api/src/paths/dwc/eml.ts]: nice catch. will address.

**Comment by cgarrettjones** [comment]
> A few things that I noticed are missing from the EML:
> 
> Project:
> 
>     * Permits
> 
>     * Location boundary description
> 
> 
> Survey:
> 
>     * Permits
> 
>     * Survey Methodology
> 
>     * Ancillary Species
> 
> 
> I can't speak to if all of those items are essential or not, mostly permits stands out to me. But everything else in the project/survey meta is there, I believe.

per Robin:
permits - no
methodology - no
ancillary species - "maybe"

**Comment by NickPhura** [comment]
> per Robin:
> permits - no
> methodology - no
> ancillary species - "maybe"

Right, no worries then

**Review by NickPhura — APPROVED** [review]
Nice work

**Pull request merged**